credentials/apps/badges/tests/test_services.py

```diff
@@ -239,6 +239,126 @@ def test_process_penalties_one_datarule_fail(self):
         process_penalties(COURSE_PASSING_EVENT, "test_username", COURSE_PASSING_DATA)
         self.assertEqual(Fulfillment.objects.filter(progress=progress).count(), 2)
 
+    def test_process_single_requirement_penalty(self):
+        requirement = BadgeRequirement.objects.create(
+            template=self.badge_template,
+            event_type=COURSE_PASSING_EVENT,
+            description="Test course passing award description",
+        )
+        DataRule.objects.create(
+            requirement=requirement,
+            data_path="course.display_name",
+            operator="ne",
+            value="B",
+        )
+        progress = BadgeProgress.objects.create(username="test_username", template=self.badge_template)
+        Fulfillment.objects.create(progress=progress, requirement=requirement)
+        self.assertEqual(Fulfillment.objects.filter(progress=progress).count(), 1)
+
+        penalty = BadgePenalty.objects.create(template=self.badge_template, event_type=COURSE_PASSING_EVENT)
+        penalty.requirements.add(requirement)
+        PenaltyDataRule.objects.create(
+            penalty=penalty,
+            data_path="course.display_name",
+            operator="eq",
+            value="A",
+        )
+        process_penalties(COURSE_PASSING_EVENT, "test_username", COURSE_PASSING_DATA)
+        self.assertEqual(Fulfillment.objects.filter(progress=progress).count(), 0)
+    
+    def test_process_one_of_grouped_requirements_penalty(self):
+        requirement_a = BadgeRequirement.objects.create(
+            template=self.badge_template,
+            event_type=COURSE_PASSING_EVENT,
+            description="Test course passing award description",
+            group="a_or_b",
+        )
+        requirement_b = BadgeRequirement.objects.create(
+            template=self.badge_template,
+            event_type=COURSE_PASSING_EVENT,
+            description="Test course passing award description",
+            group="a_or_b",
+        )
+        DataRule.objects.create(
+            requirement=requirement_a,
+            data_path="course.display_name",
```

Comment:
Please refactor to the course.course_key (since we'll hide course.display_name by default and course key is the more complicated case).